drv/medusa-seq-server/src/main.rs

```diff
@@ -0,0 +1,358 @@
+                ringbuf_entry!(if good {
+                    Trace::FrontIOBoardPhyOscGood
+                } else {
+                    Trace::FrontIOBoardPhyOscBad
+                });
+
+                front_io_board
+                    .phy()
+                    .set_osc_good(good)
+                    .map_err(MedusaError::from)
+                    .map_err(RequestError::from)
+            }
+            // The oscillator is already marked good and this state only changes
+            // if it (and by extension the whole front IO board) is power
+            // cycled. In that case the value of this register in the FPGA is
+            // automatically reset when the bitstream is loaded and the other
+            // arm of this match would be taken.
+            //
+            // So ignore this call if the oscillator has been found good since the last power
+            // cycle of the front IO board.
+            PhyOscState::Good => Ok(()),
+        }
+    }
+
+    fn reset_front_io_phy(
+        &mut self,
+        _: &RecvMessage,
+    ) -> Result<(), RequestError<MedusaError>> {
+        self.actually_reset_front_io_phy()
+            .map_err(RequestError::from)
+    }
+}
+
+impl NotificationHandler for ServerImpl {
+    fn current_notification_mask(&self) -> u32 {
+        notifications::TIMER_MASK
+    }
+
+    fn handle_notification(&mut self, _bits: u32) {
+        let next_deadline = sys_get_timer().now + TIMER_INTERVAL;
+
+        sys_set_timer(Some(next_deadline), notifications::TIMER_MASK);
+    }
+}
+
+#[export_name = "main"]
+fn main() -> ! {
+    let mut buffer = [0; idl::INCOMING_SIZE];
+
+    let mut server = ServerImpl {
+        power_control: PowerControl::new(),
+        front_io_board: None,
+    };
+
+    // Enable the front IO hot swap controller and probe for a front IO board.
+    match server.front_io_board_preinit() {
+        Ok(true) => {
+            ringbuf_entry!(Trace::FrontIOBoardPresent);
+            ringbuf_entry!(Trace::FrontIOBoardPowerGood);
+
+            let mut front_io_board = FrontIOBoard::new(
+                FRONT_IO.get_task_id(),
+                AUXFLASH.get_task_id(),
+            );
+
+            front_io_board.init().unwrap_lite();
+
+            // TODO: check/load VPD data into packrat.
+
+            // So far the front IO board looks functional. Assign it to the
+            // server, implicitly marking it present for the lifetime of this
+            // task.
+            server.front_io_board = Some(front_io_board);
+        }
+        Ok(false) => {
+            ringbuf_entry!(Trace::FrontIOBoardNotPresent);
+            server.power_control.v12_qsfp_out.set_enable(false);
+        }
+        Err(MedusaError::FrontIOBoardPowerFault) => {
+            ringbuf_entry!(Trace::FrontIOBoardPowerFault)
+        }
+        // Something went wrong getting the HSC status, eject.
+        Err(_) => panic!("unknown front IO board preinit failure"),
```

Comment:
This code is grafted over from how the `sidecar-seq-server` handles this currently and will be reworked heavily once I can land a large refactor. That being said, we actually should never reach this branch because `front_io_board_preinit` only returns a `MedusaError::FrontIOBoardPowerFault` error variant. I'll change this to an `unreachable` and update the comment accordingly.